engine/uieventmanager.cpp

```diff
@@ -15,34 +17,66 @@ void UIEventManager::ProcessEvents(const UIElementCollection& uiElements)
 
     for (auto& uiElement : uiElements)
     {
-        if (uiElement->GetType() != UIElementType::Button)
+        switch (uiElement->GetType())
         {
-            continue;
-        }
-
-        auto& button = static_cast<Button&>(*uiElement);
-        if (button.GetColliderBox()->CheckPointInBounds(mousePosition))
+        case UIElementType::Button:
```

Comment:
Okay, I chose to try out the visitor pattern. I have been wanted to try out more design patterns because I don't really see many that often (In fact this is the first time I have seen or implemented visitor.) But yeah I like it. I think it will make adding and working with these ui elements in the future a lot nicer.

Reply:
Awesome. Good works, looks like I would expect. If you want to chat over the implications of this or any other design pattern, I'm reading a very interesting book on them right now and would love to converse about them to deepen my own knowledge. I also see Design Patterns as a weakness in my coding skills.

The book is [here](https://www.amazon.com/Software-Design-Principles-Patterns-High-Quality/dp/1098113160), and so far it is SUPER useful, finally helping them click but it is dense.

Reply:
(I have also been surprised at how many design patterns I had encountered but just didn't know the proper names, so you know a lot of them, you just don't know it yet ;) )